src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -231,101 +216,163 @@ private static bool CanOptimizeSingleSpreadAsCollectionBuilderArgument(BoundColl
-                spanConstructor = ((MethodSymbol)_factory.WellKnownMember(
-                    collectionTypeKind == CollectionExpressionTypeKind.Span ? WellKnownMember.System_Span_T__ctor_Array : WellKnownMember.System_ReadOnlySpan_T__ctor_Array)!).AsMember(spanType);
+                return null;
             }
 
-            BoundExpression array;
-            if (TryOptimizeSingleSpreadToArray(node, arrayType) is { } optimizedArray)
+            BoundExpression createSpan(BoundCollectionExpression node, NamedTypeSymbol spanType, bool isReadOnlySpan)
             {
-                array = optimizedArray;
-            }
-            else if (ShouldUseKnownLength(node, out _))
-            {
-                array = CreateAndPopulateArray(node, arrayType);
-            }
-            else
-            {
-                // The array initializer has an unknown length, so we'll create an intermediate List<T> instance.
-                // https://github.com/dotnet/roslyn/issues/68785: Emit Enumerable.TryGetNonEnumeratedCount() and avoid intermediate List<T> at runtime.
-                var list = CreateAndPopulateList(node, elementType, elements);
+                Debug.Assert(isReadOnlySpan
+                    ? spanType.OriginalDefinition == (object)_compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T)
+                    : spanType.OriginalDefinition == (object)_compilation.GetWellKnownType(WellKnownType.System_Span_T));
+
+                if (tryCreateNonArrayBackedSpan(node, spanType, isReadOnlySpan) is { } spanValue)
+                    return spanValue;
 
-                Debug.Assert(list.Type is { });
-                Debug.Assert(list.Type.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_List_T), TypeCompareKind.AllIgnoreOptions));
+                var arrayType = getBackingArrayType(spanType);
+                var arrayValue = createArray(node, arrayType, targetsReadOnlyCollection: isReadOnlySpan);
 
-                var listToArray = ((MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_Collections_Generic_List_T__ToArray)).AsMember((NamedTypeSymbol)list.Type);
-                array = _factory.Call(list, listToArray);
+                var wellKnownMember = isReadOnlySpan ? WellKnownMember.System_ReadOnlySpan_T__ctor_Array : WellKnownMember.System_Span_T__ctor_Array;
+                var spanConstructor = _factory.WellKnownMethod(wellKnownMember).AsMember(spanType);
+                assertTypesAreCompatible(_compilation, arrayType, spanConstructor.Parameters[0].Type);
+                return _factory.New(spanConstructor, arrayValue);
+
+                [Conditional("DEBUG")]
+                static void assertTypesAreCompatible(CSharpCompilation compilation, TypeSymbol arrayType, TypeSymbol constructorParameterType)
+                {
+                    var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
```

Comment:
maybe doc this.  not sure what the implicit ref checkk is.

Reply:
There was an assert , verifying that types strictly matche. As it turned out, with generics we can avoid intermediate conversion to save some IL (see [this commit](https://github.com/dotnet/roslyn/pull/80104/commits/11951b4a33294794c19ed1c69b732984bc8a30f2) in particular), so the assert widened to include implicit reference conversion as well. Given that internal compiler APIs require use site info to classify a conversion, I extracted that assert into a helper, so that I can have a multiline debug-only assert helper. I believe such pattern is not uncommon in the compiler codebase